### PR TITLE
WAS-492: Extend named pipe wrapper

### PR DESCRIPTION
**Pull request by jolene-2s**
This PR applies the changes from twosense/windows-service#27 to a fork of the original named-pipe-wrapper repo, in order to see our changes more effectively and keep the wrapper as a separate package. I had to upgrade some of the test framework packages in order to get the unit tests to run, and have then applied changes in NamedPipeServer.cs and ExternalClientTests.cs. I also added a simple CI so that we can verify the build and tests are working.

### CODE REVIEW

File: UnitTests/ExternalClientTests.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            Logger.Debug("Setting up test...");
+
+            _barrier.Reset();
+
+            _server = new NamedPipeServer<byte[]>(PipeName, true);
+            _client = new NamedPipeClientStream("external_client_test_pipe");
+
+            _server.Error += ServerOnError;
+
+            _server.StartServerOnly();
```
Comment (amorta) [inline UnitTests/ExternalClientTests.cs:54]: What's difference between calling _`server.StartServerOnly();` and `_server.Start();`?
On my environment, when I call `_server.Start()` I get the following exception:
```
Named pipe is broken or disconnected: System.UnauthorizedAccessException: Access to the path is denied.
   at System.IO.Pipes.NamedPipeServerStream.Create(String pipeName, PipeDirection direction, Int32 maxNumberOfServerInstances, PipeTransmissionMode transmissionMode, PipeOptions options, Int32 inBufferSize, Int32 outBufferSize, PipeSecurity pipeSecurity, HandleInheritability inheritability, PipeAccessRights additionalAccessRights)
```

File: UnitTests/ExternalClientTests.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        private readonly ManualResetEvent _barrier = new ManualResetEvent(false);
+
+        #region Setup and teardown
+
+        [SetUp]
+        public void SetUp()
+        {
+            Logger.Debug("Setting up test...");
+
+            _barrier.Reset();
```
Comment (jtanios) [inline UnitTests/ExternalClientTests.cs:47]: What is a barrier, and why does it need to be reset?

File: NamedPipeWrapper/NamedPipeServer.cs [hunk 3]
```diff
@@ -236,13 +276,39 @@ public void Stop()
...
 			// initiate a dummy connection that will allow the thread to exit.
 			var dummyClient = new NamedPipeClient<TRead, TWrite>(_pipeName);
 			dummyClient.Start();
 			dummyClient.WaitForConnection(TimeSpan.FromSeconds(2));
 			dummyClient.Stop();
 			dummyClient.WaitForDisconnection(TimeSpan.FromSeconds(2));
+
+		}
+
+		public void Read()
```
Comment (jtanios) [inline NamedPipeWrapper/NamedPipeServer.cs:294]: Why do we need a `Read()` method? Isn't that being handled in the `ClientOnReceiveMessage()` event handler?

File: NamedPipeWrapper/NamedPipeServer.cs [hunk 4]
```diff
@@ -236,13 +276,39 @@ public void Stop()
...
+		{
+			try
+			{
+				if (_server.IsConnected && !_reader.EndOfStream)
+				{
+					string message = _reader.ReadLine();
+					if (message != null)
+					{
+						// Temporary - do something else with the command when we receive it
+						Console.WriteLine(String.Join("", message));
```
Comment (jtanios) [inline NamedPipeWrapper/NamedPipeServer.cs:304]: How would this be used by the hosting service (e.g., Windows Service)? Should it emit an event?

File: NamedPipeWrapper/NamedPipeServer.cs [hunk 5]
```diff
@@ -236,13 +276,39 @@ public void Stop()
...
+					}
+				}
+			}
+			catch (Exception e)
+			{
+				Console.WriteLine("Could not read from pipe");
```
Comment (jtanios) [inline NamedPipeWrapper/NamedPipeServer.cs:310]: Are we OK with ignoring the exception and just logging to console? Or should it raise an exception to the hosting app?

Comment (jtanios) [reply NamedPipeWrapper/NamedPipeServer.cs:310]: I see that `private void OnError(Exception exception)` has been utilized to raise read/write exceptions.

File: NamedPipeWrapper/NamedPipeServer.cs [hunk 6]
```diff
@@ -74,13 +86,25 @@ public class Server<TRead, TWrite>
...
+		private readonly NamedPipeServerStream _server;
+		private readonly StreamReader _reader;
+
 		/// <summary>
 		/// Constructs a new <c>NamedPipeServer</c> object that listens for client connections on the given <paramref name="pipeName"/>.
 		/// </summary>
 		/// <param name="pipeName">Name of the pipe to listen on</param>
-		public Server(string pipeName)
+		/// /// <param name="serverOnlyFlag">Flag indicating whether we are running this as server-only or not</param>
+		public Server(string pipeName, bool option = false)
```
Comment (jtanios) [inline NamedPipeWrapper/NamedPipeServer.cs:97]: It's unclear what the `option` parameter is used for. What would be a better name for it? Also, it should be added to the docstring.

File: NamedPipeWrapper/NamedPipeServer.cs [hunk 7]
```diff
@@ -74,13 +86,25 @@ public class Server<TRead, TWrite>
...
 		{
 			_pipeName = pipeName;
+			_option = option;
+
+			if (_option)
+			{
+				_server = new NamedPipeServerStream(pipeName, PipeDirection.InOut, 10, PipeTransmissionMode.Message);
+				_reader = new StreamReader(_server);
```
Comment (jtanios) [inline NamedPipeWrapper/NamedPipeServer.cs:105]: It looks like the `NamedPipeConnection` already supports 2-way communication. Why use the `NamedPipeServerStream`?

**Review by jtanios — COMMENTED** [review]
I didn't really dig deep into the changes and I have many questions. It would be great if we can have an overview of the changes

File: NamedPipeWrapper/NamedPipeServer.cs (hunk 7 above)
Comment (amorta) [reply NamedPipeWrapper/NamedPipeServer.cs:105]: I managed to get the 2-way communication working by looking at the [ExampleCLI](https://github.com/twosense/named-pipe-wrapper/tree/master/ExampleCLI) and modifying their sample. However, 2-way communication didn't work for me when `_option` was true.

My experiment results:

- When `_option` is **true** and server is started by calling **`Start()`** -> I get "Named pipe is broken or disconnected" error

- When `_option` is **true** and server is started by calling **`StartServerOnly()`** -> No exception, but 2-way communication doesn't work 

- When `_option` is **false** and server is started by calling **`StartServerOnly`** -> `System.NullReferenceException`

- When `_option` is **false** and server is started by calling **`Start`** ->  2-way communication works

**Pull request closed without merging**